Libraries/LibGemini/Job.h

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright (c) 2020, The SerenityOS developers.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <AK/Optional.h>
+#include <LibCore/NetworkJob.h>
+#include <LibCore/TCPSocket.h>
+#include <LibGemini/GeminiRequest.h>
+#include <LibGemini/GeminiResponse.h>
+
+namespace Gemini {
+
+class Job : public Core::NetworkJob {
+public:
+    explicit Job(const GeminiRequest&);
+    virtual ~Job() override;
+
+    virtual void start() override = 0;
+    virtual void shutdown() override = 0;
+
+    GeminiResponse* response() { return static_cast<GeminiResponse*>(Core::NetworkJob::response()); }
+    const GeminiResponse* response() const { return static_cast<const GeminiResponse*>(Core::NetworkJob::response()); }
+
+protected:
+    void finish_up();
+    void on_socket_connected();
+    virtual void register_on_ready_to_read(Function<void()>) = 0;
+    virtual void register_on_ready_to_write(Function<void()>) = 0;
+    // FIXME: I want const but Core::IODevice::can_read_line populates a cache with this
```

Comment:
Then `Core::IODevice::can_read_line()` should be `const` and make its cache `mutable`

Reply:
That's a thing?

C++: where every standard adds random things you never knew you needed, because you never wanted them.

Reply:
It could totally be a type:

```c++
template<typename T>
class Mutable {
public:
    Mutable(T value): m_value(value) {}
    T& operator *() const { return const_cast<T&>(m_value); }
private:
    T m_value;
};


// Later:
Mutable<Cache> m_some_cache;
```

But they decided to go with a keyword instead.

Reply:
Hmm...could be a nice addition; I personally didn't know the purpose of `mutable` until now
`Mutable<T>` would be quite clear about its purpose though.

Reply:
I don't think we should add a class to reimplement a language keyword :)

Reply:
shame.
Maybe next time :)

Reply:
I've opened an issue for this here - https://github.com/SerenityOS/serenity/issues/2219 - I think this deserves its own change set.